### PR TITLE
GPII-2996 gcp ci volumes

### PR DESCRIPTION
**Pull request by mrtyler**
This is a second take on the work in #80. This PR replaces #80 (and https://github.com/gpii-ops/exekube/pull/8).

Apologies for starting a new PR with new commits, abandoning previous comments and making it harder to see what you've already reviewed. With the change in strategy, I felt this would be the simplest way with the cleanest git history.

The new strategy is based on @stepanstipl's suggestion of using Docker Volumes instead of Bind Mounts. This sidesteps a lot of wacky docker and docker-compose file ownership issues, and is (I think Stepan would argue) a more Docker-ish way to do things.

This PR is not quite finished, but I've tested it on both MacOS and the CI worker. I am able to `rake apply_secret_mgmt` and `rake apply_infra` and `rake clean` and `rake clobber`. I'm posting an early version to continue to gather feedback.

Some downsides of the new strategy:
* Credentials are now managed by Docker and lack even file-level permissions. This is not an awesome situation for multi-user systems, but Docker itself is already not an awesome situation for multi-user systems. Please let me know if you can think of a different way to approach this set of problems.
   * Some of the mitigations in my response to Stepan [here](https://github.com/gpii-ops/gpii-infra/pull/80#issuecomment-413739753) might help with this.
* Rake code running outside of the container can no longer decide if certain tasks have been completed (like authenticating to gcloud or downloading `owner.json`) without running a container and checking. This adds about 1-4s to rake runtimes in my unscientific experience.
   * @natarajaya's changes in #77 to run rake inside the container might help with this?
   * I have some other ideas that I'm thinking through.

I have not yet addressed Stepan's concerns from https://github.com/gpii-ops/gpii-infra/pull/80#pullrequestreview-143176738. My next step is figuring out how to handle credentials on the CI worker, with an eye toward minimizing manual steps.

### CODE REVIEW

**Comment by stepanstipl** [comment]
It looks good imho with docker volumes.

What would be the strategy to run the same code against different organization - ie. using the same env, but different service account(s)?

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -40,46 +68,123 @@ dev:
...
     - rake clobber
 
-promote-to-stg:
+gcp-dev:
+  stage: dev
+  tags:
+    - gcp
+  script:
+    - cd gcp/live/dev
+    - rake clobber
```
Comment (stepanstipl) [inline .gitlab-ci.yml]: Why need to run `clobber` before the rake?

File: gcp/README.md [hunk 2]
```diff
@@ -46,6 +46,10 @@ Initial instructions based on [exekube's Getting Started](https://exekube.github
    * @mrtyler requested a quota bump to 100 Projects.
       * He only authorized his own email for now, to see what it did. But it's possible other Ops team members will need to go through this step.
 
+## One-time CI Setup
+1. Log in to the CI Worker and clone this repo.
+1. `cd gpii-infra/ && rake -f rakefiles/ci_save_all.rake`
```
Comment (amatas) [inline gcp/README.md]: does this rake file exist?

File: .gitlab-ci.yml (hunk 1 above)
Comment (mrtyler) [reply .gitlab-ci.yml]: Safety. Remember that `after_script` is not guaranteed to complete.

File: gcp/README.md (hunk 2 above)
Comment (mrtyler) [reply gcp/README.md]: Oops! Fixed.

**Comment by mrtyler** [comment]
> What would be the strategy to run the same code against different organization - ie. using the same env, but different service account(s)?

The strategy would be to modify this code to account for the new situation. It depends on how we manage service account credentials, but I don't see this as a big deal (copy owner2.json and owner3.json alongside owner.json? dedicated volumes for each credentials?).

Do you have a more specific concern? A specific strategy suggestion?

**Comment by stepanstipl** [comment]
I think based on today's discussion the scenarios to be covered will be:
- Run common stg with account for Test organization (test1.gpii.net)
- Run common prd with account for main (RfP) organization
- Run eatch of the gcp environments (dev, stg, prd) under the accounts for the respective environment -> each project has their own project owner svc account

So basically we will want the ability to switch between multiple gcloud service accounts (across different orgs too) for different stages, or even during stage (stg will first run common part under "common" account and then gcp part under stg account).

**Comment by mrtyler** [comment]
Sergey helped me figure out the last build gcp-dev build failure (in a manual test), so I'm going to merge this and test it For Real!

The last commit disables gcp-stg and all steps thereafter. This is because we still have Projects called "gcp-stg" and "gcp-prd" instead of the updated naming scheme "gcp-gpii-stg" and "gcp-gpii-prd", and I didn't want to touch anything while @amatas and @stepanstipl are wrapping up #60.

File: .gitlab-ci.yml (hunk 1 above)
Comment (natarajaya) [reply .gitlab-ci.yml]: I suggest we add `rake destroy_tfstate[k8s]` and `rake destroy_secrets[default]` after `rake clobber`, to forcefully repopulate secrets to make sure that part of secret management code still works as expected.

Comment (mrtyler) [reply .gitlab-ci.yml]: `clobber` is what cleans up all the Volumes, so it wants to go last. How about after `destroy`?

Comment (natarajaya) [reply .gitlab-ci.yml]: Ouch! It won't work after `clobber` in this case, we need to be still authenticated. After `destroy` is the right place.

Comment (mrtyler) [reply .gitlab-ci.yml]: https://github.com/gpii-ops/gpii-infra/pull/94

**Pull request merged**